Controller snapshots: topics

ztlpn opened the pull request:
Add topics-related data to controller snapshots.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes
* none

mmaslankaprv commented on src/v/cluster/topic_table.cc:
```diff
@@ -791,6 +793,456 @@ topic_table::apply(create_non_replicable_topic_cmd cmd, model::offset o) {
+                .last_update_finished_revision
+                = p_it->second.last_update_finished_revision});
+
+            co_await ss::coroutine::maybe_yield();
```
are we sure we can yield here ? Is it possible that topic/partition map will change during that operation ?

ztlpn replied:
should be okay, we are doing it with `mux_state_machine::_apply_mtx` locked

mmaslankaprv commented on src/v/cluster/partition_balancer_state.h:
```diff
@@ -47,6 +47,8 @@ class partition_balancer_state {
       const std::vector<model::broker_shard>& prev,
       const std::vector<model::broker_shard>& next);
 
+    ss::future<> apply_snapshot(const controller_snapshot&);
```
I am wondering why do we need to keep the balancer state in the snapshot now ? Previously the state wasn't persistent

ztlpn replied:
The balancer needs a list of ntps with violated rack constraints that is maintained by applying topic table commands. When applying a snapshot, we need to rebuild this list as well.

mmaslankaprv commented on src/v/cluster/partition_balancer_state.cc:
```diff
@@ -76,6 +77,58 @@ void partition_balancer_state::handle_ntp_update(
+            continue;
+        }
+
+        for (const auto& [p_id, partition] : topic.partitions) {
```
should we `maybe_yield` in this loop ?

ztlpn commented:
[force push](https://github.com/redpanda-data/redpanda/compare/eb06e40b2bbc2825b5c2649557c9ef732a99267b..baaa414f6ab5cfdafe6ae9fa907e91e2ee0d4d04) changes: fixed a partition allocator restore bug, added a couple of maybe_yields

dotnwat commented:
epic
